src/undo/MergeLayerDownUndoAction.cpp

```diff
@@ -0,0 +1,62 @@
+#include "MergeLayerDownUndoAction.h"
+
+#include "control/Control.h"
+#include "control/layer/LayerController.h"
+#include "gui/XournalView.h"
+#include "model/Document.h"
+#include "model/Layer.h"
+#include "model/PageRef.h"
+
+#include "i18n.h"
+
+MergeLayerDownUndoAction::MergeLayerDownUndoAction(LayerController* layerController, const PageRef& page,
+                                                   Layer* upperLayer, Layer* lowerLayer, int upperLayerPos):
+        UndoAction("MergeLayerDownUndoAction"),
+        upperLayerPos(upperLayerPos),
+        layerController(layerController),
+        upperLayer(upperLayer),
+        lowerLayer(lowerLayer),
+        upperLayerID(upperLayerPos + 1),
+        lowerLayerID(upperLayerPos) {
+    this->page = page;
+}
+
+auto MergeLayerDownUndoAction::getText() -> std::string { return _("Merge layer down"); }
+
+auto MergeLayerDownUndoAction::undo(Control* control) -> bool {
+    // remove all elements present in the upper layer from the lower layer again
+    const bool free_elems = false;  // don't free the elems, they're still used
+    for (Element* elem: this->upperLayer->getElements()) { this->lowerLayer->removeElement(elem, free_elems); }
+
+    // add the upper layer back at its old pos
+    layerController->insertLayer(this->page, this->upperLayer, upperLayerPos);
+
+    // set the selected layer back to the ID of the upper layer
+    this->page->setSelectedLayerId(this->upperLayerID);
+
+    this->undone = true;
+
+    // Rebuild the layer menu, so that the correct new layer selection is
+    // properly displayed in the sidebar. Without this, the visually displayed
+    // selection defaults to the top layer, but if the user were to delete the
+    // current layer via `shift+ctrl+l` the merged layer would disappear, as it
+    // is the one actually selected.
+    layerController->fireRebuildLayerMenu();
+
+    return true;
+}
+
+auto MergeLayerDownUndoAction::redo(Control* control) -> bool {
```

Comment:
This should not reimplement the action but use the actual implementation of the action. Otherwise, changes to the actual action might miss the implementation here. To limit the coupling of the modules, you might want to keep the implementation here and use it within the controller instead of the other way around.

Reply:
As I don't see any other of the layer actions implement it this way, I have to clarify: have the actual implementation of the action in the Undo Action's `redo()` member function and then in the Layer Controller's implementation initialize the undo action, call `redo()` to do the action initially and then give that undo action to the Undo Redo Handler?

Reply:
Yes that would be my way of doing this. It keeps the full logic of the action contained in one class and whenever you want to do or undo it you have to use the class.

Reply:
Despite the small semantic confusion of `redo` = do for the first time, I do like this. However, it does expose some of the murkyness around some of this:

I call `LayerController::fireRebuildLayerMenu` to ensure that the selection (red outline) _shown_ in the sidebar actually reflects the current selection. I call `XournalView::layerChanged` to ensure that the page is rerendered (e.g. if the layer we merge with is invisible, the merge shouid change what the page looks like, hence rerender needed).  
__However, if I call `layerChanged` _before_ `fireRebuildLayerMenu` the program freezes.__

The rough cause of the freeze is that as `fireRebuildLayerMenu` goes through the `LayerControlListener`s and prompts them each to rebuild the menu, this call to the third listener never returns.

I'd like to investigate this a bit more (as opposed to just saying "it has to be this order, else it'll freeze, who knows why"), especially because it seems like it might be related to #3290?
Probably won't have much time to before next weekend, though.

If anybody has insight on this, feel free to contribute it.

Reply:
In the implementation of the undo/redo action if we call `fireRebuildLayerMenu`
_after_ `layerChanged` it will freeze the program, e.g. if we have the regular
layer sidebar open, because upon destruction of the old previews we end up
waiting forever (I didn't quite wait this long) on `jobRunningMutex` in
`XournalScheduler::finishTask`.

### Callstack:

    LayerController::fireRebuildLayerMenu
        SidebarPreviewLayers::rebuildLayerMenu
            SidebarPreviewLayers::updatePreviews
                SidebarPreviewBaseEntry::~SidebarPreviewBaseEntry
                    XournalScheduler::removeSidebar
                        XournalScheduler::removeSource
                            XournalScheduler::finishTask

Reply:
From what I can tell, to execute the render job that ends up being added because
of the `layerChanged` call, `jobRunningMutex` is locked.
The render job however is never completed, because to execute it the document
needs to be locked, which it already has been by the undo redo handler that
calls our undo/redo action, and so we get stuck here in `Document::lock`.
Thus the layer menu rebuild waits on the job lock which waits on the document
lock which waits on the overall caller (the undo/redo handler) to be done, which
of course can't happen until the layer rebuild returns.
Thus we are deadlocked if we add this render job before rebuilding the layer menu.

Reply:
Understanding this, I'm okay with just having the calls in the 'correct' order in the undo/redo action implementation, probably with a comment warning about the order. There's really no way for me to change anything else without touching stuff that has nothing to do with the PR (besides being over my head in complexity anyways).

To stick with the pattern of implementing actions in their undo/redo actions `redo()` member, this issue might pop up again.
I'd say it stems from the fact that as I implement it for example, the undo or redo really does multiple things:
- change the document (e.g. do the layer merge)
- update the UI (e.g. trigger page rerender and update layer menu)

Maybe it would make sense to give add some kind of member `post_action()` to the undo/redo action interface that the undo/redo handler could call after having executed the action and unlocked the document again. This could then do whatever additional updating might seem necessary.

@LittleHuba or @personalizedrefrigerator thoughts?

Reply:
I followed this up with #3844